src/state/routing/utils.ts

```diff
@@ -1,70 +1,95 @@
-import { InterfaceTrade, QuoteResult, V2PoolInRoute, V3PoolInRoute } from './types'
+import { GetQuoteArgs, InterfaceTrade, QuoteResult, V2PoolInRoute, V3PoolInRoute } from './types'
 
 /**
- * Transforms a Routing API quote into an array of routes that can be used to create
- * a `Trade`.
+ * Transforms a Routing API quote into an array of routes that can be used to
+ * create a `Trade`.
  */
 export function computeRoutes(
-  currencyIn: Currency | undefined,
-  currencyOut: Currency | undefined,
-  tradeType: TradeType,
-  quoteResult: Pick<QuoteResult, 'route'> | undefined
-) {
-  if (!quoteResult || !quoteResult.route || !currencyIn || !currencyOut) return undefined
+  tokenInIsNative: boolean,
+  tokenOutIsNative: boolean,
+  quoteResult?: Pick<QuoteResult, 'route'>
+):
+  | {
+      routev3: V3Route<Currency, Currency> | null
+      routev2: V2Route<Currency, Currency> | null
+      inputAmount: CurrencyAmount<Currency>
+      outputAmount: CurrencyAmount<Currency>
+    }[]
+  | undefined {
+  if (!quoteResult || !quoteResult.route) return
 
   if (quoteResult.route.length === 0) return []
 
-  const parsedTokenIn = parseToken(quoteResult.route[0][0].tokenIn)
-  const parsedTokenOut = parseToken(quoteResult.route[0][quoteResult.route[0].length - 1].tokenOut)
+  const tokenIn = quoteResult.route[0]?.[0]?.tokenIn
+  const tokenOut = quoteResult.route[0]?.[quoteResult.route[0]?.length - 1]?.tokenOut
 
-  if (parsedTokenIn.address !== currencyIn.wrapped.address) return undefined
-  if (parsedTokenOut.address !== currencyOut.wrapped.address) return undefined
+  if (!tokenIn || !tokenOut) throw new Error('Expected both tokenIn and tokenOut to be present')
+
+  const parsedCurrencyIn = tokenInIsNative ? nativeOnChain(tokenIn.chainId) : parseToken(tokenIn)
+  const parsedCurrencyOut = tokenOutIsNative ? nativeOnChain(tokenOut.chainId) : parseToken(tokenOut)
 
   try {
     return quoteResult.route.map((route) => {
       if (route.length === 0) {
         throw new Error('Expected route to have at least one pair or pool')
       }
-      const rawAmountIn = route[0].amountIn
-      const rawAmountOut = route[route.length - 1].amountOut
+      const rawAmountIn = route[0]?.amountIn
+      const rawAmountOut = route[route.length - 1]?.amountOut
 
       if (!rawAmountIn || !rawAmountOut) {
         throw new Error('Expected both amountIn and amountOut to be present')
       }
 
+      const isOnlyV2 = isV2OnlyRoute(route)
+      const isOnlyV3 = isV3OnlyRoute(route)
+
       return {
-        routev3: isV3Route(route) ? new V3Route(route.map(parsePool), currencyIn, currencyOut) : null,
-        routev2: !isV3Route(route) ? new V2Route(route.map(parsePair), currencyIn, currencyOut) : null,
-        inputAmount: CurrencyAmount.fromRawAmount(currencyIn, rawAmountIn),
-        outputAmount: CurrencyAmount.fromRawAmount(currencyOut, rawAmountOut),
+        routev3: isOnlyV3 ? new V3Route(route.map(parsePool), parsedCurrencyIn, parsedCurrencyOut) : null,
```

Comment:
nit: I prefer these booleans inlined, because they're only used once.

Reply:
ah sorry they are copy pasted from the mobile stack where we have mixed routes support, and they're used in the mixed routes logic, which is coming up very soon so will keep these in